### PR TITLE
Improved gmres

### PR DESCRIPTION
**Pull request by alewis**
Adds an improved version of Jax gmres. The new version builds the QR factorization of the Arnoldi overlap matrix by updating it at each Arnoldi iteration. This allows the solution residual to also be estimated at each Arnoldi iteration, so that the Arnoldi process can be terminated early if convergence is reached. Also adds improved docstrings and tests.

### CODE REVIEW

File: tensornetwork/backends/jax/jitted_functions.py [hunk 1]
```diff
@@ -543,8 +562,8 @@ def gmres_m(A_mv: Callable, A_args: Sequence,
     where x and b are length-n vectors, using the method of
     Generalized Minimum RESiduals with M iterations per restart (GMRES_M).
 
-    Args:
-
+    ARGS
+    ----
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:566]: Leave it as `Args:`

File: tensornetwork/backends/jax/jitted_functions.py [hunk 2]
```diff
@@ -567,54 +586,400 @@ def gmres_m(A_mv: Callable, A_args: Sequence,
     n_iter (int)    : Number of iterations at termination.
     converged (bool): Whether the desired tolerance was achieved.
     """
+    num_krylov_vectors = min(num_krylov_vectors, b.size)
     x = x0
-    converged = False
-    r, beta = gmres_residual(A_mv, A_args, b, x)
     b_norm = jnp.linalg.norm(b)
+    tol = max(tol*b_norm, atol)
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:592]: Nit: We add a space before and after `*`. Surprised linter didn't notice...

File: tensornetwork/backends/jax/jitted_functions.py [hunk 3]
```diff
@@ -567,54 +586,400 @@ def gmres_m(A_mv: Callable, A_args: Sequence,
...
+    b        : The `b` in `A @ x = b`.
+    x        : Initial guess solution.
+    tol      : Solution tolerance to achieve,
+    num_krylov_vectors : Size of the Krylov space to build.
+
+    RETURNS
+    -------
+    done     : Whether convergence was achieved.
+    beta     : Magnitude of residual (i.e. the error estimate).
+    x        : The approximate solution.
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:621]: Please use argument formatting to be consistent with the rest of the library.

Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions.py:621]: fair enough. Is it just ARGS, RESULTS -> Args: Results:?

Comment (Thenerdstation) [reply tensornetwork/backends/jax/jitted_functions.py:621]: Yeah, and the spacing. Take a look at any other function to see an example. I might add docs check in the future.

Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions.py]: Ok, should be fixed.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 4]
```diff
@@ -518,20 +520,37 @@ def implicitly_restarted_arnoldi_method(
...
   Returns:
-    gmres: A function performing gmres_m as described below.
+  -------
+  functions: A dictionary of functions:
+    functions["gmres_m"] = gmres_m
+    functions["gmres_residual"] = gmres_residual
+    functions["gmres_krylov"] = gmres_krylov
+    functions["_gs_step"] = _gs_step
+    functions["kth_arnoldi_step"] = kth_arnoldi_step
+    functions["givens_rotation"] = givens_rotation
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:546]: I don't like that this returns a dictionary. Perhaps use a NamedTuple instead?https://docs.python.org/3/library/collections.html#collections.namedtuple

File: tensornetwork/backends/jax/jitted_functions.py [hunk 5]
```diff
@@ -544,77 +563,383 @@ def gmres_m(A_mv: Callable, A_args: Sequence,
...
+    sn = -v2 / t
+    return cs, sn
+
+
+  functions["gmres_m"] = gmres_m
+  functions["gmres_residual"] = gmres_residual
+  functions["gmres_krylov"] = gmres_krylov
+  functions["_gs_step"] = _gs_step
+  functions["kth_arnoldi_step"] = kth_arnoldi_step
+  functions["givens_rotation"] = givens_rotation
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:944]: Use NamedTuple here.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 6]
```diff
@@ -518,20 +523,37 @@ def implicitly_restarted_arnoldi_method(
...
+  functions: A namedtuple of functions:
+    functions.gmres_m = gmres_m
+    functions.gmres_residual = gmres_residual
+    functions.gmres_krylov = gmres_krylov
+    functions.gs_step = _gs_step
+    functions.kth_arnoldi_step = kth_arnoldi_step
+    functions.givens_rotation = givens_rotation
   """
   jnp = jax.numpy
+  functions = dict()
```
Comment (Thenerdstation) [inline tensornetwork/backends/jax/jitted_functions.py:552]: You can delete this line.

**Comment by alewis** [comment]
I'm unable to reproduce the current build error on my machine, and the error message from Travis is pretty cryptic. I do get a typing error from block_sparse/tensordot_test.py and a linter error from tn_kera/test_layer, but I didn't change those things.

**Comment by alewis** [comment]
@Thenerdstation @mganahl Does anyone understand what is happening here?

**Comment by Thenerdstation** [comment]
Wut

**Comment by Thenerdstation** [comment]
Looks like you've hit a pytype bug. Ouch.

I think it's about time we've upgraded our pytype anyway. Let me do that tomorrow and then try this again.

**Comment by mganahl** [comment]
@alewis is this the pylint bug you hit earlier?

**Comment by mganahl** [comment]
@googlebot I consent.

**Pull request merged**